src/main/scala/nl.knaw.dans.easy.deposit/WorkerThread.scala

```diff
@@ -0,0 +1,129 @@
+import java.nio.file.{ WatchEvent, WatchKey }
+
+import better.files.File
+import nl.knaw.dans.easy.deposit.docs.StateInfo.State
+import nl.knaw.dans.lib.logging.DebugEnhancedLogging
+import nl.knaw.dans.lib.error._
+
+import scala.util.Try
+
+object WorkerThread extends DebugEnhancedLogging {
+
+  val stagedBaseDir = File("target")
+
+  def main(args: Array[String]): Unit = {
+
+    // TODO create watchService as first action in EasyDepositApiService.start
+    //  close it as last action in EasyDepositApiService.stop
+    stagedBaseDir.watchService.apply { watchService =>
+      stagedBaseDir.register(watchService, Seq(ENTRY_CREATE, ENTRY_DELETE))
+      handleAbortedByHardShutdown() // between register and take of watchService!
+      while ( { // TODO in a sub-thread (possibly with handleAbortedByHardShutdown, for a quick start)
+        // https://howtodoinjava.com/java8/java-8-watchservice-api-tutorial/
+        // take, because polling skips staged-upload-dirs created and deleted within a single cycle
+        Try(watchService.take())
+          .map(handleEvents)
+          .doIfFailure { case e =>
+            // does not get here on SIGINT despite InterruptedException thrown by take
+            // TODO would it get a ClosedWatchServiceException when the parent thread is the culprit?
+            logger.error(s"terminating ${ e.getMessage }")
+          }
+          .getOrElse(false)
+      }) {}
+    }
+  }
+
+  private def handleEvents(watchKey: WatchKey): Boolean = {
+    watchKey.pollEvents.stream.forEach(handleEvent(_))
+    watchKey.reset() // next pending event
+  }
+
+  private def handleEvent(event: WatchEvent[_]): Unit = {
+    val stagedDir = stagedBaseDir / event.context().toString
+    (event.kind(), getDepositState(stagedDir.name)) match {
+      case (ENTRY_CREATE, State.finalizing) => finalizeSubmit(stagedDir)
+      case (ENTRY_DELETE, State.draft) => calculateShas(stagedDir)
+      case _ => // either completed finalizeSubmit or started upload
+    }
+  }
+
+  private def handleAbortedByHardShutdown(): Unit = {
+    stagedBaseDir.list.foreach { abortedDir =>
+      getDepositState(abortedDir.name) match {
+        case State.draft => abortedUpload(abortedDir)
+        case State.finalizing => abortedSubmit(abortedDir)
+        case state => logger.error(s"Staged garbage : $state $abortedDir")
+      }
+    }
+    logger.info(s"So far what was found at startup in $stagedBaseDir")
+  }
+
+  private def abortedSubmit(stagedSubmitDir: File): Unit = {
+    logger.info(s"Found submit aborted by hard shutdown: $stagedSubmitDir")
+    deleteContent(stagedSubmitDir) // clean up for a fresh submit
+    finalizeSubmit(stagedSubmitDir)
+  }
+
+  def deleteContent(dir: File): Unit = {
+    dir.list.foreach { f =>
+      println(s"deleting $f")
+      if (f.isDirectory) deleteContent(f)
+      f.delete()
+    }
+  }
+
+  private def abortedUpload(stagedUploadDir: File): Unit = {
```

Comment:
Can you now simple recursively remove `stagedUploadDir`?

Reply:
I don't know of a delete method with such an option.

For submits I should keep the directory but drop its content. For uploads I need to drop the directory with its content. Other content than multipart files might indicate some problem, so I left it this way once I implemented deleteContent.

Reply:
Doesn't have`better.files` such an option? If not, maybe just use commons-io: https://commons.apache.org/proper/commons-io/javadocs/api-release/index.html